### PR TITLE
Multivariate capabilities

### PR DESCRIPTION
**Pull request by James-Large**
Simple/minimal implementation of the multivariate capabilities + some testing changes for it

see here https://github.com/sktime/sktime-dl/pull/34/files#diff-5832d822f3c64be59b410587982d269dR98, and regressor code is same 

Does NOT check for unequal length series. Don't know if there's a standard sktime method of dealing with this, but in my opinion this is an open question dependant on epxeriental aims and the particular data

Could add in zero padding to longest dimension length for example, but for data with extreme differences in length it's just not reasonable. I'd argue for the user correcting this during data loading stages rather than use checking for/correcting it within fit

Based on these factors though, the docstrings likely need updating for clarification to be honest. But that's not a job for today 

Otherwise, multivariate classification testing turned on with basicmotions as the test data. 

In the interest of testing time, commented out the all-network univariate testing since it is covered indirectly by the high level and multivarate tests

Regressor/forecasting tests reduced to 10 epochs of training too

Lastly, snuck in a fix of updating the regressor training metrics from accuracy to mse

### CODE REVIEW

File: sktime_dl/deeplearning/base/estimators/_classifier.py [hunk 1]
```diff
@@ -93,16 +93,25 @@ def check_and_clean_data(self, X, y=None, input_checks=True):
...
-                    "Input should either be a 2d numpy array, or a pandas dataframe with a single column of Series objects (networks cannot yet handle multivariate problems")
+            if X.shape[1] > 1:
+                # we have multiple columns
+                # if cells contain series, this is a multidimensional problem
+                # else if cells contain single values, this is a univariate problem with values long columns
+                #      this situation can happen with e.g. forecasting-reduced-to-regression strategies
+                if isinstance(X.iloc[0, 0], pd.Series):
+                    # todo investigate incorporating the reshaping into the data extraction instead of this 2-stage process
+                    X = np.array([[X.iloc[r, c].values for c in range(len(X.columns))] for r in range(len(X))])
+                    X = X.reshape(X.shape[0], X.shape[2], X.shape[1])  # go from [n][d][m] to [n][m][d]
```
Comment (mloning) [inline sktime_dl/deeplearning/base/estimators/_classifier.py]: I'd turn this routine into a utility function, it may also be covered already by the [tabularise](https://github.com/alan-turing-institute/sktime/blob/master/sktime/utils/data_container.py) helper function in sktime, can also be re-used in the classifier file below

File: sktime_dl/deeplearning/base/estimators/_classifier.py [hunk 2]
```diff
@@ -93,16 +93,25 @@ def check_and_clean_data(self, X, y=None, input_checks=True):
...
         if isinstance(X, pd.DataFrame):
-            if X.shape[1] > 1 or not isinstance(X.iloc[0, 0], pd.Series):
-                raise TypeError(
-                    "Input should either be a 2d numpy array, or a pandas dataframe with a single column of Series objects (networks cannot yet handle multivariate problems")
+            if X.shape[1] > 1:
+                # we have multiple columns
+                # if cells contain series, this is a multidimensional problem
+                # else if cells contain single values, this is a univariate problem with values long columns
+                #      this situation can happen with e.g. forecasting-reduced-to-regression strategies
+                if isinstance(X.iloc[0, 0], pd.Series):
```
Comment (mloning) [inline sktime_dl/deeplearning/base/estimators/_classifier.py]: I'd make this check also into a function, that'll make it easier later one if we change the data container, i.e. something like
```python
def is_nested_dataframe(X):
    return isinstance(X.iloc[0, 0], pd.Series)
```

**Comment by James-Large** [comment]
Answering both, have made a utils._data file, moved check_and_clean_data to there

Aliased a few operations, tabularise was a neater way to do the nested univariate case, but was wrong format for others. This could definitely be revisited in the future

File: sktime_dl/deeplearning/base/estimators/_regressor.py [hunk 3]
```diff
@@ -55,35 +51,42 @@ def predict(self, X, input_checks=True, **kwargs):
...
+    #             else:
+    #                 X = X.to_numpy()
+    #         else:
+    #             X = np.asarray([a.values for a in X.iloc[:, 0]])
+    #
+    #     if len(X.shape) == 2:
+    #         # add a dimension to make it multivariate with one dimension
+    #         X = X.reshape(X.shape[0], X.shape[1], 1)  # go from [n][m] to [n][m][d=1]
+    #
+    #     return X
```
Comment (Withington) [inline sktime_dl/deeplearning/base/estimators/_regressor.py]: You could remove this commented out function.

File: sktime_dl/deeplearning/cnn/_regressor.py [hunk 4]
```diff
@@ -90,7 +91,7 @@ def build_model(self, input_shape, **kwargs):
 
         model = keras.models.Model(inputs=input_layer, outputs=output_layer)
         model.compile(loss='mean_squared_error', optimizer=keras.optimizers.Adam(),
-                      metrics=['accuracy'])
+                      metrics=['mean_squared_error'])
```
Comment (Withington) [inline sktime_dl/deeplearning/cnn/_regressor.py]: Well spotted, thanks.

File: sktime_dl/deeplearning/mlp/_regressor.py [hunk 5]
```diff
@@ -98,7 +99,12 @@ def fit(self, X, y, input_checks=True, **kwargs):
         -------
         self : object
         """
-        X = self.check_and_clean_data(X, y, input_checks=input_checks)
+
+        print(X.shape, X.iloc[0,0].shape)
+        X = check_and_clean_data(X, y, input_checks=input_checks)
+        print(X.shape)
+
```
Comment (Withington) [inline sktime_dl/deeplearning/mlp/_regressor.py]: Remove the print statements?

File: sktime_dl/deeplearning/tests/test_networkcapabilities.py [hunk 6]
```diff
@@ -1,4 +1,7 @@
-from sktime.datasets import load_italy_power_demand  # , load_basic_motions
+import pandas as pd
+import numpy as np
+
```
Comment (Withington) [inline sktime_dl/deeplearning/tests/test_networkcapabilities.py]: Remove pd and np imports?

**Comment by Withington** [comment]
Good stuff. I put a couple of suggestions against lines in the code.

File: sktime_dl/deeplearning/cnn/_regressor.py (hunk 4 above)
Comment (mloning) [reply sktime_dl/deeplearning/cnn/_regressor.py]: Apparently it still works with the accuracy though ... anyway, perhaps we should make the loss also a argument of the constructor?

Comment (Withington) [reply sktime_dl/deeplearning/cnn/_regressor.py]: Nothing is being done with metrics here so it wasn't a problem.

Loss could be passed as an argument, yes, the user might like to try different loss functions.

**Comment by James-Large** [comment]
Ah, had those cleanups locally but didn't push them on Friday. Will merge once checks done
